Libplanet/Blockchain/Policies/BlockPolicy.cs

```diff
@@ -66,134 +64,141 @@ public BlockPolicy(TimeSpan blockInterval)
-                        (block.PreviousHash?.ToString() ?? "nothing")
-                    );
-                }
+            for (var i = 0; i < blocks.Count; i++)
+            {
+                Block<T> block = blocks[i];
+                InvalidBlockException exception = ValidateBlockToAppend(
+                    i, secondLastBlock, lastBlock, block);
 
-                if (block.Timestamp < prevTimestamp)
+                if (!(exception is null))
                 {
-                    return new InvalidBlockTimestampException(
-                        $"the block #{i}'s timestamp ({block.Timestamp}) is " +
-                        $"earlier than the block #{i - 1}'s ({prevTimestamp})"
-                    );
+                    return exception;
                 }
 
-                prevHash = block.Hash;
-                prevTimestamp = block.Timestamp;
+                secondLastBlock = lastBlock;
+                lastBlock = block;
             }
 
             return null;
         }
 
+        /// <inheritdoc/>
+        public InvalidBlockException ValidateBlockToAppend(
+            IReadOnlyList<Block<T>> blocks,
+            Block<T> blockToAppend)
+        {
+            int blockCount = blocks.Count;
+            Block<T> secondLastBlock = blockCount >= 2
+                ? blocks[blockCount - 2]
+                : null;
+            Block<T> lastBlock = blockCount >= 1
+                ? blocks[blockCount - 1]
+                : null;
+
+            return ValidateBlockToAppend(
+                blocks.Count,
+                secondLastBlock,
+                lastBlock,
+                blockToAppend);
+        }
+
         /// <inheritdoc />
         public int GetNextBlockDifficulty(IReadOnlyList<Block<T>> blocks)
         {
-            return ExpectDifficulties(blocks, yieldNext: true)
-                .First(t => t.Block == null)
-                .Difficulty;
+            int blockCount = blocks.Count;
+
+            if (blockCount <= 1)
+            {
+                return blockCount;
+            }
+
+            Block<T> prevPrevBlock = blocks[blockCount - 2];
+            Block<T> prevBlock = blocks[blockCount - 1];
+
+            return GetNextDifficultyFromPrevTimestamp(
+                prevPrevBlock.Timestamp,
+                prevBlock.Timestamp,
+                prevBlock.Difficulty);
         }
 
-        private IEnumerable<DifficultyExpectation> ExpectDifficulties(
-            IEnumerable<Block<T>> blocks, bool yieldNext = false)
+        private InvalidBlockException ValidateBlockToAppend(
+            int blockCount,
+            Block<T> secondLastBlock,
+            Block<T> lastBlock,
+            Block<T> blockToAppend)
         {
-            DateTimeOffset? prevTimestamp = null;
-            DateTimeOffset? prevPrevTimestamp = null;
+            int index = blockCount;
+            int difficulty = 0;
 
-            if (yieldNext)
+            if (lastBlock is Block<T> block)
```

Comment:
This test may seem a bit confusing.

```suggestion
            if (!(lastBlock is null))
```